### DISCUSSION TITLE
Handling errors thrown by Next's compiled and bundled version of `node-fetch`

### DISCUSSION THREAD

**Discussion by philwolstenholme** [vercel/next.js#46972] [Help] [answered]
### Summary

I'm trying to catch the custom `NodeFetch` errors thrown by the version of `node-fetch` that Next bundles up in `node_modules/next/dist/compiled/node-fetch/index.js`.

A `NodeFetch` error looks like this:

``` 
{
  message: 'invalid json response body at https://example.com/dodgy-syntax reason: Unexpected token < in JSON at position 0',
  type: 'invalid-json'
}
```

I'm using TypeScript, so before I can check `error.type` to see how to handle the error I need to check the instance of the error. Normally I'd do something like:

```
if (error instanceof FetchError && error.type === 'invalid-json') {
      throw new Error(`JSON parsing error: ${error.message}`);
    }
```

…and import `FetchError` from `node-fetch`, but because Next's version of `node-fetch` is a precompiled bundle of code I can't do that, I don't think?

Does anyone have any tips or should I just accept the errors provided by `node-fetch` rather than try and augment them?

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.3.0: Thu Jan  5 20:53:49 PST 2023; root:xnu-8792.81.2~2/RELEASE_X86_64
    Binaries:
      Node: 18.14.0
      npm: 9.6.0
      Yarn: N/A
      pnpm: 7.29.0
    Relevant packages:
      next: 13.2.3
      eslint-config-next: 13.2.3
      react: 18.2.0
      react-dom: 18.2.0
```

```
export interface HttpError extends Error {
  status: number;
  response: Response;
}

export class HttpError extends Error {
  constructor(response: Response) {
    super(
      `Request to ${response.url} failed with status ${response.status} (${response.statusText})`
    );
    Object.setPrototypeOf(this, HttpError.prototype);
    this.name = 'HttpError';
    this.status = response.status;
    this.response = response;
  }
}

export async function getJson(url: string, options?: RequestInit) {
  const response = await fetch(url, options);

  // @TODO: Add a proper logger to use across whole app
  console.log(`network.ts > get:`, `Fetching ${url}`);

  if (!response.ok) {
    // This should activate the closest `error.js` Error Boundary, except
    // it doesn't seem to be working at the moment. Maybe a Next 13 issue?
    throw new HttpError(response);
  }

  return response.json().catch(error => {
    console.log(error);

    if (error instanceof FetchError && error.type === 'invalid-json') { // Cannot find name 'FetchError'. Unsafe member access .type on an `any` value.eslint@typescript-eslint/no-unsafe-member-access
      throw new Error(`JSON parsing error: ${error.message}`); // Invalid type "any" of template literal expression.eslint@typescript-eslint/restrict-template-expressions
Unsafe member access .message on an `any` value.eslint@typescript-eslint/no-unsafe-member-access
    }
  });
}

```
```


### Example

_No response_

**Comment by icyJoseph** [comment, marked as answer]
If you are using Node 18, then I think Next.js doesn't polyfill fetch... 

- https://github.com/vercel/next.js/blob/canary/packages/next/src/server/node-polyfill-fetch.ts

```js
// TODO: Remove use of `any` type.
// Polyfill fetch() in the Node.js environment

if (!(global as any).fetch) {
```

Do you see the Node.js warning when running your project?

> ExperimentalWarning: The Fetch API is an experimental feature. This feature could change at any time

**Reply by philwolstenholme** [reply]
Ah, that might explain why fetch was behaving differently on the Next Dev server than it was in Jest… I use `next/jest` and import `import next from 'next';` it in my Jest setup file, so I thought everything might be in sync, but perhaps I've done something wrong

**Reply by philwolstenholme** [reply]
Although I don't see the Node.js warning… 🤔

**Reply by icyJoseph** [reply]
Yeah, it's like, Jest makes up its own environment. https://github.com/vercel/next.js/issues/43676, so it blissfully ignores your runtime fetch's... https://github.com/facebook/jest/pull/12642 but maybe it is patched? what Jest version do you have?

**Reply by icyJoseph** [reply]
A few more links:

- https://github.com/facebook/jest/issues/13834


And a solution? https://github.com/jsdom/jsdom/issues/1724#issuecomment-1446858041

**Reply by philwolstenholme** [reply]
Oh man, I hadn't even got to considering how `jsdom` was going to make this harder, I was still looking into just the Next side of things :D 

I'm on Jest and `jest-environment-jsdom` 29.5.0. I've got `require('next');` at the top of my `jest.setup.ts` file, without that I get `fetch is not defined` ReferenceErrors.

**Reply by icyJoseph** [reply]
Yeah so that'd be because either jest or jsdom, plocks it out, I generally wouldn't want to jump into implementation details of these libraries though... 

Often people just throw in `whatwg-fetch` to "get it to work", but maybe that last link I shared, where they "fix" the JSDOM environment, is what we need? (Also trying to get jest to work with a few other projects) -

I must say though that `Vitest` doesn't do this, I have some vitest tests that use fetch from my runtime.

**Reply by philwolstenholme** [reply]
Ah this is interesting. When I wrote this ticket I had `const nextJest = require('next/jest')` in my Jest setup file (a mistake) and that _was_ causing Next's bundled version of `node-fetch` to be used, but when I replace that with `require('next');` in my Jest setup file it _doesn't_ use the bundled version of `node-fetch`.

I confirmed this by adding some `console.log` calls to the `if` statement in `node_modules/next/dist/server/node-polyfill-fetch.js`.

I'll do some more investigation but this feels better now that I spotted the mistake of requiring `next/jest` instead of just `next`.

**Reply by philwolstenholme** [reply]
Haha, okay, this is getting so weird. I took a closer look at my Jest logs after lunch and it looks like both branches of the `if` statement in `node-polyfill-fetch.js` are being hit:

```
 PASS  src/utils/formatPrice.test.ts
  ● Console

    console.log
      node-polyfill-fetch.js: global.fetch is not defined, so we are polyfilling it

      at Object.<anonymous> (node_modules/next/src/server/node-polyfill-fetch.ts:5:3)

    console.log
      node-polyfill-fetch.js: global.fetch is defined, so we are not polyfilling it

      at Object.<anonymous> (node_modules/next/dist/server/node-polyfill-fetch.js:60:9)
```

I'll keep documenting my findings here in case I end up finding a solution or writing a bug report, but feel free to turn off notifications @icyJoseph so I don't keep spamming you! :D

**Reply by philwolstenholme** [reply]
I also added a `console.log('node-fetch is loaded')` to `node_modules/next/dist/compiled/node-fetch/index.js` and a similar one in `node_modules/next/dist/compiled/undici/index.js` just in case. I also console logged before the `require('next')` in `jest.setup.ts` and the `const nextJest = require('next/jest');` in `jest.config.js`.

The `next/jest` require seems to bring in `node-fetch`, even on Node v18.14.0:

```
❯ npm run test:unit

> next-playground-app-directory@0.1.1 test:unit
> jest --ci --cacheDirectory .jestcache --maxWorkers=1

Loading next/jest
node-fetch is loaded
info  - Loaded env from /Users/wolsph/Projects/shop-next/.env.test
info  - Loaded env from /Users/wolsph/Projects/shop-next/.env
warn  - You have enabled experimental feature (appDir) in next.config.js.
warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use at your own risk.
info  - Thank you for testing `appDir` please leave your feedback at https://nextjs.link/app-feedback
```

When we get to running the actual tests, `node-fetch` is being loaded too. Unlike the previous logs I shared earlier, this time it looks like `node-fetch` is being continuously brought in as a polyfill:

```
 PASS  src/utils/network.test.ts
  ● Console

    console.log
      Loading next

      at Object.log (jest.setup.ts:4:9)

    console.log
      node-polyfill-fetch.js: global.fetch is not defined, so we are polyfilling it

      at Object.<anonymous> (node_modules/next/src/server/node-polyfill-fetch.ts:5:3)

    console.log
      node-fetch is loaded

      at Object.<anonymous> (node_modules/next/dist/compiled/node-fetch/index.js:1:29)
```

And I can see errors being thrown by `node-fetch`:

```
  ● network utilities › getJson › throws a JSON parse error if response is not JSON

    expect(received).rejects.toThrow(expected)

    Expected substring: "Request to https://example.com/dodgy-syntax failed with status 200 (OK) but invalid JSON"
    Received message:   "invalid json response body at https://example.com/dodgy-syntax reason: Unexpected token < in JSON at position 0"

    at node_modules/next/dist/compiled/node-fetch/index.js:2:51220
```

**Reply by philwolstenholme** [reply]
Using https://github.com/jsdom/jsdom/issues/1724#issuecomment-1446858041 (but also adding `this.global.Request = Request;`) has meant that now `undici` is now throwing errors for my deliberate Syntax error causing test instead of `node-fetch`, this feels promising:

```
  console.log
    Loading next

      at Object.log (jest.setup.ts:4:9)

  console.log
    SyntaxError: Unexpected token < in JSON at position 0
        at JSON.parse (<anonymous>)
        at packageData (node:internal/deps/undici/undici:6403:23)
        at specConsumeBody (node:internal/deps/undici/undici:6381:14)
        at processTicksAndRejections (node:internal/process/task_queues:95:5)
        at Object.<anonymous> (/Users/wolsph/Projects/shop-next/src/utils/network.test.ts:64:7)
```

**Reply by philwolstenholme** [reply]
Yep – creating a custom jsdom environment and using `require('next')` in `jest.setup.ts` sorted out my issues :)
